### ISSUE TITLE
My Issues and Watched Issues return 400: Bad Request

### ISSUE THREAD

**Issue by brandenbyers** [steyep/alfred-jira#41]
As per #38, I am receiving an error when attempting to access My Issues or Watched Issues. However, search appears to be working.

```
[2017-03-09 10:46:05][input.scriptfilter] <?xml version="1.0"?>
<items>
  <item valid="no">
    <title>400: Bad Request</title>
  </item>
</items>
```

Navigating to `{ my jira configured url }/rest/api/2/search?jql=assignee=currentUser()` in a browser and I receive a 220,315 character JSON response.

Let me know what else I can do to help figure this one out. Even just with search, you have created a great workflow!

**Comment by steyep** [comment]
@brandenbyers can you open up the settings pane and confirm that you have *at least* one status enabled under the **Enabled Issue Statuses** heading (enabled statuses have a blue background)? If you don't see any statuses under that heading, what do you get if you run this command from the `alfred-jira` directory?
```
echo "require('./lib/jira').getStatuses().then(console.log).catch(console.error);" | node
```

**Comment by brandenbyers** [comment]
Yes, I have about 100 Enabled Issue Statuses selected in blue. Most of those statuses are from projects that aren't mine. Not sure if it helps to troubleshoot, but there are 150+ Projects and I have most of those disabled except for 5 of them (but I haven't disabled any of the Issue Statuses yet).

**Comment by steyep** [comment]
Can you paste the below script into the terminal and run it from the `alfred-jira` directory? It's going to make a GET request to the jira instance and (presumably) throw a `400` error but it's also going to make the same request via the browser. If you are able to see the JSON dump in the browser, verify that the address in the browser starts with the same protocol that's printed in the terminal. Then if you don't mind replying with the results of the script (everything below the line of asterisks). 

If the script is successful, that may be indicative of an issue with Alfred's access to the credentials stored in your keychain. For peace of mind, can you confirm that you have an entry in `Keychain Access.app` for `alfred-jira`?
```
node << 'EOF'
const auth = require('./lib/jira/auth');
const config = require('./lib/jira/config');
const request = require('axios');
const sh = require('child_process').exec;
const getAvailableStatuses = () => config.options.available_issues_statuses.join('", "');
const sort = () => config.sort.map(s => s.name + '+' + (s.desc ? 'DESC' : 'ASC')).join(',+');
const searchInProjects = () => {
  let options = config.options || {}, projects = options.available_projects || [];
  return projects.length ? '+AND+project+in+("' + projects.join('", "') + '")' : '';
}
auth.setConfig()
  .then(() => {
    let query = config.url + 'rest/api/2/search?jql=assignee=currentUser()'
          + searchInProjects() 
          + '+AND+status+in+("' + getAvailableStatuses() + '")'
          + '+order+by+' + sort();
    return request.get(query, config.req);
  })
  .then(res => {
    let totalIssues = res.data.total;
    console.log('REQUEST SUCCESSFUL!')
    console.log(totalIssues + ' issues found');
  })
  .catch(res => {
    let url = res.config.url;
    let message = res.response.statusText;
    sh(`open '${url}'`);
    console.log(`
********************************************
STATUS: ${res.response.status}
MESSAGE: ${message}
PROTOCOL: ${url.match(/https?:\/\//)}
ERROR: 
${res.response.data.errorMessages}
`, res.response.data.errors)});
EOF
```

**Comment by brandenbyers** [comment]
This is the json response that I get from that script:

```
{
"errorMessages": [
"Error in the JQL Query: '\\Q' is an illegal JQL escape sequence. The valid escape sequences are \\', \\\", \\t, \\n, \\r, \\\\, '\\ ' and \\uXXXX. (line 1, character 298)"
],
"errors": {}
}
```

And I have confirmed that I do have a keychain entry for `alfred-jira`.

**Comment by steyep** [comment]
Interesting. I wonder if the query is simply too long with all of your projects enabled. The [JIRA API Reference](https://docs.atlassian.com/jira/REST/cloud/#api/2/search-searchUsingSearchRequest) suggests that could be a possibility: 

> GET vs POST: If the JQL query is too large to be encoded as a query param you should instead POST to this resource.

From what I've read, it seems like the query limit may be around 2000 characters. I have roughly 110 projects enabled but my query string only ends up being 1831 characters. Can you run the following script to verify that posting the request resolves your issue and that your query is longer than 2000 characters?

```
node << 'EOF'
const auth = require('./lib/jira/auth');
const config = require('./lib/jira/config');
const request = require('axios');
const sh = require('child_process').exec;
const getAvailableStatuses = () => config.options.available_issues_statuses.join('", "');
const sort = () => config.sort.map(s => s.name + '+' + (s.desc ? 'DESC' : 'ASC')).join(',+');
const searchInProjects = () => {
  let options = config.options || {}, projects = options.available_projects || [];
  return projects.length ? '+AND+project+in+("' + projects.join('", "') + '")' : '';
}
auth.setConfig()
  .then(() => {
    let query = 'rest/api/2/search?jql=assignee=currentUser()'
          + searchInProjects() 
          + '+AND+status+in+("' + getAvailableStatuses() + '")'
          + '+order+by+' + sort();
    console.log(`\n*********************************\nQUERY LENGTH: ${query.length}`)
    let data = { jql: query.replace(/^[^=]+=/, '').replace(/\+/g, ' ') };
    return request.post(config.url + 'rest/api/2/search', data, config.req);
  })
  .then(res => {
    let totalIssues = res.data.total;
    console.log('REQUEST SUCCESSFUL!')
    console.log(totalIssues + ' issues found');
  })
  .catch(res => {
    let url = res.config.url;
    let message = res.response.statusText;
    sh(`open '${url}'`);
    console.log(`
STATUS: ${res.response.status}
MESSAGE: ${message}
PROTOCOL: ${url.match(/https?:\/\//)}
ERROR: 
${res.response.data.errorMessages}
`, res.response.data.errors)});
EOF
```

**Comment by brandenbyers** [comment]
It appears that my query length is less than yours.

```
*********************************
QUERY LENGTH: 1718

STATUS: 400
MESSAGE: Bad Request
PROTOCOL: https://
ERROR:
Error in the JQL Query: '\Q' is an illegal JQL escape sequence. The valid escape sequences are \', \", \t, \n, \r, \\, '\ ' and \uXXXX. (line 1, character 298)
 {}
```

Also, it may be worth noting that although there are 150+ projects in my Jira org, I only have 5 of them enabled in the workflow settings. However, the previous script you had me run included all of them.

**Comment by steyep** [comment]
Okay so it appears that it's not an issue of length. Can you share the query string so that can I can try to make sense of the "illegal JQL escape sequence" error that's getting thrown at character 298? 

```
node << 'EOF'
const auth = require('./lib/jira/auth');
const config = require('./lib/jira/config');
const getAvailableStatuses = () => config.options.available_issues_statuses.join('", "');
const sort = () => config.sort.map(s => s.name + '+' + (s.desc ? 'DESC' : 'ASC')).join(',+');
const searchInProjects = () => {
  let options = config.options || {}, projects = options.available_projects || [];
  return projects.length ? '+AND+project+in+("' + projects.join('", "') + '")' : '';
}
auth.setConfig().then(() => {
  let query = 'rest/api/2/search?jql=assignee=currentUser()' + searchInProjects() + '+AND+status+in+("' + getAvailableStatuses() + '")' + '+order+by+' + sort();
  console.log(query);
})
EOF
```

**Comment by brandenbyers** [comment]
This is what is returned: `rest/api/2/search?jql=assignee=currentUser()+AND+project+in+("SHIA", "PCDAND", "PCDIOS", "PCDWEBK", "PURE")+AND+status+in+("Pending Approval", "In Progress", "Reopened", "Change Closed", "Closed", "To Do", "Done", "Ready for Code Review", "Ready for QA", "Abandoned", "Blocked", "Needs External Review", "In Development\QA", "Pending Functional Requirements", "In Development", "Pending Backend", "Testing", "P.O. Acceptance", "Pending Requirements / Triage", "Researching", "Awaiting Approval", "In Backlog", "Abandon", "Ready to work", "Epics", "Pending Requirements", "Complete", "In QA", "pre-backlog", "In Test", "Approved", "Rejected", "Queued for Dev", "Ready", "Backlogged", "Code Complete", "Accepted", "On Hold", "Selected for Development", "Review", "Merged", "Testing Triage", "Prioritized Backlog", "Release Manager Approval", "Change Implemented", "PostMortem", "Executive Approval", "In Review", "Reject", "Finished", "Start", "Incoming", "Pending Department Approvals", "Reviewing Request", "Triage", "Future", "Near Term", "Verified", "Ready for Automation", "In Automation", "QA Verified", "Verifying in Jenkins Job", "Shelved", "Shelved In Progress", "Automation", "Created policy", "Ready for Development", "Release Review", "Backlog", "Released", "Inbox", "Requires Triage", "Support Review", "Verify", "Limited Release", "Prospect", "Active Customer", "Roadmap", "Blockers", "Icebox", "In Testing", "Declined", "Resolved", "QA Platform Backlog", "Automation Backlog", "In Manual Verification", "Waiting for Review", "Investigation", "Integration Testing", "Waiting on 3rd Party", "In Progress - Unit Tests", "Deployed to Production", "Requirements Gathering")+order+by+Created+DESC,+Priority+DESC`

**Comment by steyep** [comment]
Awesome! I will look into escaping the escape character (`\`) in queries. But, for what it's worth, `Requirements / Triage` shouldn't cause a problem. I have several statuses that contain a forward slash.

**Comment by steyep** [comment]
@brandenbyers I have pushed up a fixed for this (among other things) in the `feature/42-generic-query-bookmarks` branch. If you have a few moments, could you check it out and let me know if that resolves your issue?

**Comment by Lightheaded** [comment]
@steyep, I have the same issue related to escaping characters. For me the troublesome status was "Monitoring & controlling" and I tried feature/42-generic-query-bookmarks, which unfortunately did not resolve the issue. I am able to use the workflow when I disable the said filter though.

Also, when running npm build, answering a lower-case "y" to Re-install doesn't seem to do anything. A minor thing, but maybe you want to check for both Y and y. Sorry for littering :(

**Comment by brandenbyers** [comment]
Sorry for the delay. I tested `feature/42-generic-query-bookmarks` with the `In Development\QA` enabled but it still errors. Again, it isn't a problem anymore for me given that I don't need that status enabled anyway. But if you need me to test anything else, I can do so.

**Comment by steyep** [comment]
Okay, guys, I've been trying my hardest to get the same errors you're getting. Any query I create with a `\` or `&` is properly escaped so I'm thinking it has something to do with the fact that those special characters are in a status. Those statuses are written and read from the config file (`~/.alfred-jira/config.json`) and I'm curious if the `JSON.parse`/`JSON.stringify` methods are doing something strange to the data **or** if the response from the server is returning a URI encoded string.

Would either of you mind opening that file and letting me know what value is saved for that status? Or you could run this command:
```
grep -e '"name": \(".*In Development.*"\|".*Monitoring.*".\)' ~/.alfred-jira/config.json
```

The latest commit that I've pushed up to `feature/42-generic-query-bookmarks` *should* handle the latter encoding scenario but, like I said, I'm flying blind 😎 

@Lightheaded, the capitalized confirmation in the installation script is actually by design, so that someone won't accidentally overwrite something they didn't intend to (that's also why the "Y" is capitalized in the `Y/n` line). That said, I'm not *opposed* to changing it – just thought it was common convention.

**Comment by brandenbyers** [comment]
Here is what I get from the config.json:

`"name": "In Development\\QA", "name": "In Development"`

**Comment by Lightheaded** [comment]
In my case it's

    "name": "Monitoring & controlling",

This is probably not helping. I will try to capture the request and see what is actually fired against Jira when I have a little more time.

**Comment by steyep** [comment]
I appreciate both of you taking the time to test that for me. Judging from the values in your `config.json` I'm going to assume that pulling down the most recent changes to the `feature/42-generic-query-bookmarks` branch didn't resolve the issues either?

**Comment by danieljimenez** [comment]
What I found with mine was that it was trying to reference deleted (or archived?) projects. Not sure, I'm not a JIRA pro by any means we just started using it.

**Comment by steyep** [comment]
I recently stood up a local test instance of Jira, created workflows containing the two problematic statuses mentioned in this issue (`Monitoring & controlling` and `In Development\QA`), and am happy to report that I was able to reproduce the problem. 

I've pushed a bugfix to the `bugfix/41-special-characters-in-query` branch that resolved the issue for me and plan to include it in the `v1.0.8` release.

**Comment by richardmuhler** [comment]
Hey, I just tried setting up the plugin and I am still experiencing this issue. Has this been resolved? I am running 1.0.8.

**Comment by steyep** [comment]
@richardmuhler can you checkout the `develop` (or `bugfix/41-special-characters-in-query`) branch and see if that fixes your problem? I haven't merged those into `master` yet.

**Comment by richardmuhler** [comment]
I switched to develop:
```
~/GitHub/alfred-jira develop*
❯ git checkout develop
> Already on 'develop'
Your branch is up-to-date with 'origin/develop'.
```
Then cleared cache and logged in again and still get the 400 error when listing my tasks.
I can search for some of my tasks but far from all.

**Comment by earlonrails** [comment]
bugfix/41-special-characters-in-query this branch works for me fixing this error thanks! Awesome workflow!

**Comment by steyep** [comment]
Glad to hear it! Thanks for the feedback
